browser/brave_rewards/rewards_tab_helper.cc

```diff
@@ -192,13 +205,43 @@ void RewardsTabHelper::MaybeSavePublisherInfo() {
   }
 
   // The Rewards system currently assumes that the |publisher_info| table is
-  // populated by calling `GetPublisherActivityFromUrl` as the user nativates
+  // populated by calling `GetPublisherActivityFromUrl` as the user navigates
   // the web. Previously, this was accomplished within the background script of
   // the Rewards extension.
   rewards_service_->GetPublisherActivityFromUrl(
       tab_id_.id(), GetWebContents().GetLastCommittedURL().spec(), "", "");
 }
 
+void RewardsTabHelper::OnCreatorDetected(
+    std::optional<CreatorDetectionScriptInjector::Result> result) {
+  if (!rewards_service_ || !result) {
+    return;
+  }
+
+  SetPublisherIdForTab(result->id);
+
+  if (!result->id.empty()) {
+    // When a creator has been detected for the current tab, we must send the
+    // creator data to the utility process so that the "publisher_info" database
+    // table can be populated. We must also notify the utility process that a
+    // "page view" has started for this creator.
+    auto visit = mojom::VisitData::New();
+    visit->tab_id = static_cast<uint32_t>(tab_id_.id());
+    visit->domain = result->id;
+    visit->name = result->name;
+    visit->path = "";
+    visit->url = result->url;
+    visit->favicon_url = result->image_url;
+    if (auto platform = GetMediaPlatformFromPublisherId(result->id)) {
+      visit->provider = *platform;
+    }
+
+    rewards_service_->GetPublisherActivityFromUrl(visit->Clone());
```

Comment:
this should really be encapsulated in something like `UpdateCreatorForTab`. I have no idea what is going on in these three calls and this also makes it hard to test vs unit testing a single method on RewardsService. We really shouldn't be exposing these kinds of low level calls in the service.

Reply:
side note: why is this called `GetPublisherActivityFromUrl` if it doesn't actually return anything? Also it doesn't take a URL as a param which is confusing. Separately from this PR this method name should probably change.

Reply:
oh, I guess an overloaded version of `GetPublisherActivityFromUrl` was added in this PR so the name should at least change to `GetPublisherActivityFromVisitData` and then change `GetPublisherActivityFrom` for both in a separate PR

Reply:
Yes - these methods are pretty ancient and the names reflect that. I've updated the new method name as suggested, and I'll add an issue to fix the names up more generally.

Reply:
what about my comment about wrapping these three method calls in a single call to the service? If multiple actions need to be taken in response to something it's best for the service to handle that, not the caller. Although this is not a user action, it should still follow the same principle of having a single call to the service for a given event that occurs outside the component

Reply:
IMO, the `RewardsService` API isn't great, but refactoring `RewardsService` is more of a separate concern.

Reply:
I'm not suggesting refactoring the entire class, but we can certainly simplify this specific case, no? Just because other code is problematic doesn't mean we should keep doing the same thing when we could at least wrap these calls. It's only adding a couple of methods to RewardsService to move the details inside of it

Reply:
Filed a follow-up issue here: https://github.com/brave/brave-browser/issues/41832 to address the RewardsService AC-related API issues. It will be more clear what the API should be after the upcoming AC changes have been finalized.

Reply:
fyi I made a similar comment in this PR to make things more unit testable https://github.com/brave/brave-core/pull/26082/files#r1813441440

Reply:
unit testing tab helpers is not very practical, but unit testing the service is

Reply:
I won't block on this as long as we have good test coverage for everything that gets called here, but given that this is new code being added I'm struggling with why we can't just add a new method to the service to wrap these three calls so we can unit test it.